datafusion/core/src/physical_plan/topk/mod.rs

```diff
@@ -0,0 +1,516 @@
+            // If there was a previous minimum value, decrement its use
+            if let Some(prev_min) = self.inner.pop() {
+                if prev_min.batch_id == batch_entry.id {
+                    batch_entry.uses -= 1;
+                } else {
+                    self.store.unuse(prev_min.batch_id);
+                }
+                // update memory accounting
+                let prev_size = owned_row_size(&prev_min.row);
+                assert!(self.owned_row_bytes >= prev_size);
+                self.owned_row_bytes -= prev_size;
+            }
+        }
+    }
+
+    /// Returns the values stored in this heap, from values low to high, as a single
+    /// [`RecordBatch`]
+    pub fn emit(&self, schema: SchemaRef) -> Result<RecordBatch> {
+        // Indicies for each row within its respective RecordBatch
+        let indicies: Vec<_> = self
+            .inner
+            .iter()
+            .enumerate()
+            .map(|(i, k)| (i, k.index))
+            .collect();
+
+        let num_columns = {
+            let Some(first_value) = self.inner.get(0) else {
+                return Ok(RecordBatch::new_empty(schema));
+            };
+            self.store
+                .get(first_value.batch_id)
+                .expect("invalid batch id")
+                .batch
+                .num_columns()
+        };
+
+        // build the output columns one at time, using the
+        // `interleave` kernel to pick rows from different arrays
+        let output_columns: Vec<_> = (0..num_columns)
+            .map(|col| {
+                let input_arrays: Vec<_> = self
+                    .inner
+                    .iter()
+                    .map(|k| {
+                        let entry =
+                            self.store.get(k.batch_id).expect("invalid stored batch id");
+                        entry.batch.column(col) as &dyn Array
+                    })
+                    .collect();
+
+                // at this point `indices` contains indexes within the
+                // rows and `input_arrays` contains a reference to the
+                // relevant Array for that index. `interleave` pulls
+                // them together into a single new array
+                Ok(interleave(&input_arrays, &indicies)?)
+            })
+            .collect::<Result<_>>()?;
+
+        Ok(RecordBatch::try_new(schema, output_columns)?)
+    }
+
+    /// return the size of memory used by this heap, in bytes
+    fn size(&self) -> usize {
+        std::mem::size_of::<Self>()
+            + (self.inner.capacity() * std::mem::size_of::<TopKRow>())
+            + self.store.size()
+            + self.owned_row_bytes
+    }
+}
+
+/// Size of memory owned by `row` until row::size() is available
+/// TODO file upstream ticket in arrow-rs to add this
+fn owned_row_size(row: &OwnedRow) -> usize {
+    std::mem::size_of_val(row) + row.as_ref().len() // underlying data, doesn't account for capacity
```

Comment:
FWIW OwnedRow is an exact sized container, so there is no capacity to account for